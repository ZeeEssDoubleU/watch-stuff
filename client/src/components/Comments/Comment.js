import React, { useState } from "react";
import { Button, Image } from "semantic-ui-react";

import { getFormattedTimeAgo } from "../../utils/format-time";

import "./Comment.scss";
import Ratings from "../Ratings/Ratings";

const Comment = props => {
	const [collapsed, toggleCollapse] = useState(true);

	const commentId = props.comment.id;
	const comment = props.comment.snippet.topLevelComment.snippet;
	const commentIcon = comment.authorProfileImageUrl;
	const commentAuthor = comment.authorDisplayName;
	const commentAuthorUrl = comment.authorChannelUrl;
	const commentText = <p className="comment-text">{comment.textOriginal}</p>;
	const commentDate = getFormattedTimeAgo(comment.updatedAt);
	const commentLikes = comment.likeCount;

	const commentVisibility = collapsed ? "collapsed" : "expanded";
	const buttonText = collapsed ? "Read more" : "Read less";

	return (
		<div className="comment-container">
			<a href={commentAuthorUrl} target="_blank">
				<Image className="user-image" src={commentIcon} circular />
			</a>
			<div className="comment-items">
				<div className="user-name">{commentAuthor}</div>
				<div className="comment-date">{commentDate}</div>
				<div className="comment-body">
					<div className={commentVisibility}>{commentText}</div>
					<Button
						className="show-more-button"
						compact
						onClick={() => toggleCollapse(!collapsed)}>
						{buttonText}
					</Button>
				</div>
				<div className="comment-actions">
					<Ratings
						likes={commentLikes}
						isComment={true}
						commentId={commentId}
					/>
					<Button size="mini" compact>
						REPLY
					</Button>
				</div>
			</div>
		</div>
	);
};

export default Comment;
