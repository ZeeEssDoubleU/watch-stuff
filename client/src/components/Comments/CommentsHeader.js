import React from "react";
import PropTypes from "prop-types";
import { Button, Icon } from "semantic-ui-react";

import "./CommentsHeader.scss";

const CommentsHeader = props => {
	const commentsCount = props.commentsCount
		? Number(props.commentsCount).toLocaleString()
		: props.commentsCount;
	return (
		<div className="comments-header">
			<h4>{commentsCount} Comments</h4>
			<Button
				className="sort-button"
				basic
				compact
				icon
				labelPosition="left">
				<Icon name="align left" />
				SORT BY
			</Button>
		</div>
	);
};

CommentsHeader.propTypes = {};

export default CommentsHeader;
