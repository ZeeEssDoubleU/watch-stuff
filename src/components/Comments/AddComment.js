import React from "react";
import { Image, Form, TextArea } from "semantic-ui-react";

import "./AddComment.scss";

const AddComment = props => (
	<div className="add-comment">
		<Image
			className="user-image"
			src="https://via.placeholder.com/48x48"
			circular
		/>
		<Form>
			<TextArea
				autoHeight
				rows={1}
				placeholder="Add a public comment..."
			/>
		</Form>
	</div>
);

export default AddComment;
