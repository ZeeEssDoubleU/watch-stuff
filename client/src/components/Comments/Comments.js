import React, { useState, useEffect } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import "./Comments.scss";
import CommentsHeader from "./CommentsHeader";
import AddComment from "./AddComment";
import Comment from "./Comment";
import InfiniteScroll from "../InfiniteScroll/InfiniteScroll";

import {
	selector_commentsLoaded,
	selector_commentsNPT,
	selector_commentsByVideo,
} from "../../store/reducers/comments";

// TODO - 1.5*
const Comments = props => {
	// initialize .comments elem state and variables to determine comments infinite scroll height
	const [commentsHeight, setCommentsHeight] = useState(0);
	const commentsElem = document.querySelector(".comments");
	const commentsElemHeight = commentsElem
		? window.innerHeight - commentsElem.offsetTop
		: window.innerHeight;
	useEffect(() => {
		setCommentsHeight(commentsElemHeight);
	}, [commentsElemHeight]);

	// variables and effect to fetch more related videos based on visibilty of loader
	const loader = document.querySelector(".comments > .loader-container");
	const loaderOffset = loader ? loader.offsetTop : 0;
	const loaderVisible = loaderOffset < window.innerHeight;
	useEffect(() => {
		if (props.commentsLoaded && loaderVisible) {
			props.fetchComments(props.videoId, props.commentsNPT);
		}
	}, [loaderOffset]);

	// fetchMoreComments functions used in InfiniteScroll
	const fetchMoreComments = () => {
		if (props.commentsLoaded && props.commentsNPT) {
			props.fetchComments(props.videoId, props.commentsNPT, 10);
		}
	};

	const comments = props.comments
		? props.comments.map(comment => (
				<Comment comment={comment} key={comment.id} />
		  ))
		: null;

	return (
		<div className="comments-container">
			<CommentsHeader
				className="comments-header"
				commentsCount={props.commentsCount}
			/>
			<AddComment />
			<div className="comments" style={{ height: commentsHeight }}>
				<InfiniteScroll
					bottomReachedCallback={() => fetchMoreComments()}
					showLoader={props.commentsNPT ? true : false}>
					{comments}
				</InfiniteScroll>
			</div>
		</div>
	);
};

const mapStateToProps = (state, props) => ({
	comments: selector_commentsByVideo(state, props.videoId),
	commentsLoaded: selector_commentsLoaded(state, props.videoId),
	commentsNPT: selector_commentsNPT(state, props.videoId),
});

export default withRouter(
	connect(
		mapStateToProps,
		null,
	)(Comments),
);
