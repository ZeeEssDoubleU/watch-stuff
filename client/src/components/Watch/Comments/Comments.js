import React, { Component } from "react";
import PropTypes from "prop-types";

import "./Comments.scss";
import CommentsHeader from "./CommentsHeader/CommentsHeader";
import AddComment from "./AddComment/AddComment";
import Comment from "./Comment/Comment";

class Comments extends Component {
	render() {
		return (
			<div className="comments">
				<CommentsHeader />
				<AddComment />
				<Comment />
				<Comment />
				<Comment />
				<Comment />
				<Comment />
				<Comment />
				<Comment />
				<Comment />
			</div>
		);
	}
}

Comments.propTypes = {};

export default Comments;
