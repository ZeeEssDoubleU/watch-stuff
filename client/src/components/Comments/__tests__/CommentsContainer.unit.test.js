import React from "react";
import { shallow } from "enzyme";

import Comments from "../Comments";

describe("Comments", () => {
	test("renders without props", () => {
		const wrapper = shallow(<Comments />);
		expect(wrapper).toMatchSnapshot();
	});

	test("renders with commentCount", () => {
		const wrapper = shallow(<Comments commentCount={112499} />);
		expect(wrapper).toMatchSnapshot();
	});
});
