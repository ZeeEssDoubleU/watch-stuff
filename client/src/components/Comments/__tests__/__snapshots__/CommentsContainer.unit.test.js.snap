// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments renders with commentCount 1`] = `
<div
  className="comments-container"
>
  <CommentsHeader
    className="comments-header"
    commentCount={402}
  />
  <AddComment />
  <Comment />
  <Comment />
  <Comment />
  <Comment />
  <Comment />
  <Comment />
  <Comment />
  <Comment />
</div>
`;

exports[`Comments renders without props 1`] = `
<div
  className="comments-container"
>
  <CommentsHeader
    className="comments-header"
    commentCount={402}
  />
  <AddComment />
  <Comment />
  <Comment />
  <Comment />
  <Comment />
  <Comment />
  <Comment />
  <Comment />
  <Comment />
</div>
`;
